Base/QTGUI/qSlicerScriptedIOOptionsWidget.cxx

```diff
@@ -0,0 +1,169 @@
+QString qSlicerScriptedIOOptionsWidget::pythonSource()const
+{
+  Q_D(const qSlicerScriptedIOOptionsWidget);
+  return d->PythonSource;
+}
+
+//-----------------------------------------------------------------------------
+bool qSlicerScriptedIOOptionsWidget::setPythonSource(const QString& newPythonSource, const QString& _className, bool missingClassIsExpected)
+{
+  Q_D(qSlicerScriptedIOOptionsWidget);
+
+  if (!Py_IsInitialized())
+    {
+    return false;
+    }
+
+  if(!newPythonSource.endsWith(".py") && !newPythonSource.endsWith(".pyc"))
+    {
+    return false;
+    }
+
+  // Extract moduleName from the provided filename
+  QString moduleName = QFileInfo(newPythonSource).baseName();
+
+  QString className = _className;
+  if (className.isEmpty())
+    {
+    className = moduleName;
+    if (!moduleName.endsWith("IOOptionsWidget"))
+      {
+      className.append("IOOptionsWidget");
+      }
+    }
+
+  d->PythonCppAPI.setObjectName(className);
+  d->PythonClassName = className;
+
+  // Get a reference (or create if needed) the <moduleName> python module
+  PyObject * module = PyImport_AddModule(moduleName.toUtf8());
+
+  // Get a reference to the python module class to instantiate
+  PythonQtObjectPtr classToInstantiate;
+  if (PyObject_HasAttrString(module, className.toUtf8()))
+    {
+    classToInstantiate.setNewRef(PyObject_GetAttrString(module, className.toUtf8()));
+    }
+  else if (missingClassIsExpected)
+    {
+    // Class is not defined for this object, but this is expected, not an error
+    return false;
+    }
+
+  if (!classToInstantiate)
+    {
+    PythonQt::self()->handleError();
+    PyErr_SetString(PyExc_RuntimeError,
+                    QString("qSlicerScriptedIOOptionsWidget::setPythonSource - "
+                            "Failed to load scripted IOOptions Widget: "
+                            "class %1 was not found in file %2").arg(className).arg(newPythonSource).toUtf8());
+    return false;
+    }
+
+  PyObject* self = d->PythonCppAPI.instantiateClass(this, className, classToInstantiate);
+  if (!self)
+    {
+    return false;
+    }
+
+  d->PythonSource = newPythonSource;
+
+  return true;
+}
+
+//-----------------------------------------------------------------------------
+PyObject* qSlicerScriptedIOOptionsWidget::self() const
+{
+  Q_D(const qSlicerScriptedIOOptionsWidget);
+  return d->PythonCppAPI.pythonSelf();
+}
+
+//-----------------------------------------------------------------------------
+void qSlicerScriptedIOOptionsWidget::updateGUI(const qSlicerIO::IOProperties& ioProperties)
+{
+  Q_D(qSlicerScriptedIOOptionsWidget);
+  PythonQtObjectPtr arguments;
+  arguments.setNewRef(PythonQtConv::QVariantMapToPyObject(ioProperties));
+  PyObject* result = d->PythonCppAPI.callMethod(d->UpdateGUIMethod, arguments);
+  Py_DECREF(arguments);
+  if (!result)
+    {
+    // Method call failed (probably an omitted function), call default implementation
```

Comment:
I know that we do this all the time, but now that you are working on this and you are so knowledgable about Python integration, I would ask: is it possible to do this better? The main limitation of this is that if an uncaught exception is raised in the reimplemented method then it is silently ignored and it is assumed that the method is just not overridden.

Reply:
There is way to address this and simplified across Slicer, will do so in a follow-up PR.